In [1]:
import os 
import subprocess as sp
import pandas as pd
import glob 
import seaborn as sns 
import matplotlib.pyplot as plt 

os.chdir('/mnt/BioHome/jreyna/jreyna/projects/dchallenge/')

## Number of eQTL before filtering (eQTL Catalog)

In [2]:
before_data = []
before_glob = 'results/main/eqtl/*/ge/*_ge_*.all.prefilter.num_eqtls.txt'
before_glob = glob.glob(before_glob)

In [3]:
before_data = []
for fn in before_glob:
    with open(fn) as fr:
        
        eqtl, ge = os.path.basename(fn).split('_ge_')
        ge = ge.split('.')[0]
        
        num_eqtls = int(fr.read().strip())
        
        before_data.append([eqtl, ge, num_eqtls])

In [4]:
before_df = pd.DataFrame(before_data)
before_df.columns = ['eqtl', 'ge', 'num_eqtls_pre']

## Number of eQTL post filtering (eQTL Catalog)

In [5]:
post_data = []
post_glob = 'results/main/eqtl/*/ge/*_ge_*.all.postfilter.num_eqtls.txt'
post_glob = glob.glob(post_glob)

In [6]:
post_data = []
for fn in post_glob:
    with open(fn) as fr:
        
        eqtl, ge = os.path.basename(fn).split('_ge_')
        ge = ge.split('.')[0]
        
        num_eqtls = int(fr.read().strip())
        
        post_data.append([eqtl, ge, num_eqtls])

In [7]:
post_df = pd.DataFrame(post_data)
post_df.columns = ['eqtl', 'ge', 'num_eqtls_post']

In [8]:
post_df

eqtl                    ge  num_eqtls_post
0               GTEx    brain_hypothalamus          423929
1               GTEx         brain_putamen          559636
2               GTEx       artery_coronary          551959
3               GTEx         kidney_cortex          123311
4               GTEx  heart_left_ventricle         1091506
..               ...                   ...             ...
72    Schmiedel_2018         NK-cell_naive          472002
73           GENCORD                   LCL          823148
74           GENCORD            fibroblast          575290
75           GENCORD                T-cell          716949
76  van_de_Bunt_2015      pancreatic_islet          178434

[77 rows x 3 columns]

## Number of eQTL post filtering (Mu et al., 2021)

In [9]:
mu_data = []
mu_glob = 'results/main/2021_Nikhil_eQTL/Data/eqtl_sqtl_summ_stats/*_eQTL/*.input.txt'
mu_glob = glob.glob(mu_glob)

In [10]:
mu_data = []
for fn in mu_glob:
    with open(fn) as fr:
                
        eqtl, ge = fn.split('/')[-2:]
        eqtl = eqtl.split('_')[0]
        ge = ge.split('.')[0]
       
        num_eqtls = len(fr.readlines())
        
        mu_data.append([eqtl, ge, num_eqtls])

In [11]:
mu_df = pd.DataFrame(mu_data)
mu_df.columns = ['eqtl', 'ge', 'num_eqtls_mu']

# initializing with the original name 
cell_dict = {k:v for k, v in zip(mu_df['ge'], mu_df['ge'])}

# change the cell line name for those cell lines I care about 
replace_dict = {'Neutrophil': 'neutrophil',
             'Monocyte': 'monocyte',
             'TREG_NAIVE': 'Treg_naive',
             'TREG_MEMORY': 'Treg_memory',
             'B_NAIVE': 'B-cell_naive',
             'TFH': 'Tfh_memory',
             'TH1-17': 'Th1-17_memory',
             'CD4_N_STIM': 'CD4_N_STIM',
             'NK_CD16POS': 'NK-cell_naive',
             'NONCLASSICAL_MONOCYTES': 'monocyte_CD16_naive',
             'CD8_NAIVE': 'CD8_T-cell_naive',
             'TH17': 'Th17_memory',
             'CD8_N_STIM': 'CD8_T-cell_anti-CD3-CD28',
             'CLASSICAL_MONOCYTES': 'monocyte_naive',
             'CD4_NAIVE': 'CD4_T-cell_naive	',
             'TH1': 'Th1_memory',
             'TH2': 'Th2_memory'}
cell_dict.update(replace_dict)
mu_df['ge'] = mu_df['ge'].map(cell_dict, )

In [30]:
# replace the eqtl source name when necessary
eqtl_sources = {k: v for k, v in zip(mu_df['eqtl'], mu_df['eqtl'])}
eqtl_sources['DICE'] = 'Schmiedel_2018'
mu_df['eqtl'] = mu_df['eqtl'].map(eqtl_sources)

In [31]:
mu_df.head()

eqtl            ge  num_eqtls_mu
0        GEUVADIS           CEU          9294
1        GEUVADIS           YRI          2474
2  Schmiedel_2018    Treg_naive          2106
3  Schmiedel_2018   Treg_memory          1972
4  Schmiedel_2018  B-cell_naive          1827

## Compare the eQTL Catalog before and after FDR filtering

In [42]:
all_df = pd.merge(before_df, post_df, on=['eqtl', 'ge'], how='outer')
all_df['%Post'] = all_df['num_eqtls_post'] / all_df['num_eqtls_pre'] * 100

In [43]:
pd.set_option('display.max_rows', 100) 
display(all_df)
pd.set_option('display.max_rows', 20) 

eqtl                               ge  num_eqtls_pre  \
0               GTEx                          stomach      164513595   
1               GTEx                    brain_caudate      175054708   
2               GTEx                           muscle      134303457   
3               GTEx                brain_spinal_cord      168668568   
4               GTEx                 brain_cerebellum      167769402   
5               GTEx                  artery_coronary      162448099   
6               GTEx                           breast      172228883   
7               GTEx                            ovary      163412345   
8               GTEx                 colon_transverse      168613256   
9               GTEx                           vagina      167971169   
10              GTEx             heart_left_ventricle      154646246   
11              GTEx               brain_hypothalamus      177038437   
12              GTEx                           uterus      163759505   
13              GTEx                            liver      145722181   
14              GTEx                   brain_amygdala      172424838   
15              GTEx                    colon_sigmoid      163844032   
16              GTEx                            blood      117815659   
17              GTEx                 skin_sun_exposed      163322320   
18              GTEx      brain_cerebellar_hemisphere      162510726   
19              GTEx                    adrenal_gland      159291457   
20              GTEx             esophagus_muscularis      158515838   
21              GTEx                     brain_cortex      172395548   
22              GTEx                        pituitary      175697076   
23              GTEx                       fibroblast      136548606   
24              GTEx                              LCL      134305450   
25              GTEx           heart_atrial_appendage      157561336   
26              GTEx          brain_nucleus_accumbens      174567689   
27              GTEx                    esophagus_gej      159952566   
28              GTEx           brain_substantia_nigra      169669525   
29              GTEx                             lung      170356939   
30              GTEx                     artery_aorta      155343364   
31              GTEx                           testis      217040861   
32              GTEx                         pancreas      134927983   
33              GTEx                    kidney_cortex      166674968   
34              GTEx                 esophagus_mucosa      151397457   
35              GTEx                    artery_tibial      149696926   
36              GTEx  brain_anterior_cingulate_cortex      169679980   
37              GTEx             brain_frontal_cortex      168439512   
38              GTEx                         prostate      178134566   
39              GTEx                    brain_putamen      170637762   
40              GTEx                 adipose_visceral      163519153   
41              GTEx             skin_not_sun_exposed      163743853   
42              GTEx             minor_salivary_gland      160186457   
43              GTEx             adipose_subcutaneous      161593005   
44              GTEx                brain_hippocampus      177968032   
45              GTEx                           spleen      160676960   
46              GTEx                     nerve_tibial      167790232   
47              GTEx                          thyroid      171339394   
48              GTEx                  small_intestine      172495178   
49        Lepik_2017                            blood      103124065   
50        Quach_2016                    monocyte_R848      168940282   
51        Quach_2016                monocyte_Pam3CSK4      171429490   
52        Quach_2016                     monocyte_IAV      188522612   
53        Quach_2016                   monocyte_naive      177145991   
54        Quach_2016                     monocyte_LPS   

In [45]:
summary = all_df.drop('%Post', axis=1).describe()

In [46]:
summary

num_eqtls_pre  num_eqtls_post
count   7.700000e+01    7.700000e+01
mean    1.578687e+08    9.954745e+05
std     1.958584e+07    6.194735e+05
min     8.660671e+07    1.233110e+05
25%     1.513975e+08    5.486080e+05
50%     1.615930e+08    7.828380e+05
75%     1.696695e+08    1.299706e+06
max     2.170409e+08    2.946020e+06

## Add Mu et al eQTL's to compare

In [47]:
all_plus_mu = all_df.merge(mu_df, on=['eqtl', 'ge'])

In [51]:
all_plus_mu['%Mu'] = all_plus_mu['num_eqtls_mu'] / all_plus_mu['num_eqtls_pre'] * 100

In [52]:
all_plus_mu

eqtl                        ge  num_eqtls_pre  num_eqtls_post  \
0        BLUEPRINT                  monocyte      107504217         2213353   
1        BLUEPRINT                    T-cell      114006812         1818085   
2        BLUEPRINT                neutrophil       86606708         1838013   
3   Schmiedel_2018               Treg_memory      155297785          588584   
4   Schmiedel_2018            monocyte_naive      138656372          643412   
5   Schmiedel_2018                Treg_naive      156779364          623104   
6   Schmiedel_2018       monocyte_CD16_naive      139733495          614987   
7   Schmiedel_2018                Tfh_memory      155294143          603596   
8   Schmiedel_2018             NK-cell_naive      156385911          472002   
9   Schmiedel_2018          CD8_T-cell_naive      155723260          706670   
10  Schmiedel_2018               Th17_memory      151190959          701417   
11  Schmiedel_2018              B-cell_naive      154773799          633355   
12  Schmiedel_2018                Th1_memory      153424768          513914   
13  Schmiedel_2018             Th1-17_memory      150672748          636433   
14  Schmiedel_2018                Th2_memory      154397233          615587   
15  Schmiedel_2018  CD8_T-cell_anti-CD3-CD28      141347873          510059   

       %Post  num_eqtls_mu       %Mu  
0   2.058852          7717  0.007178  
1   1.594716          7212  0.006326  
2   2.122252          6400  0.007390  
3   0.379003          1972  0.001270  
4   0.464033          2253  0.001625  
5   0.397440          2106  0.001343  
6   0.440114          2105  0.001506  
7   0.388679          2054  0.001323  
8   0.301819          1883  0.001204  
9   0.453799          2374  0.001524  
10  0.463928          2521  0.001667  
11  0.409213          1827  0.001180  
12  0.334962          1903  0.001240  
13  0.422394          2188  0.001452  
14  0.398703          2161  0.001400  
15  0.360854          2002  0.001416

What is causing the difference between the Mu et al 2021 data and the eQTL Catalogue? In the next statements I'm considering BLUEPRINT monocytes. The eQTL Catalog we see 107,504,217 eQTL tests whereas for Mu et al we see 7,717 tests. Even after filtering the eQTL Catalogue for FDR < 0.05 there are 2,213,353 eQTL tests.   

#### Questions
- Difference in the FDR?

#### Answers
- Mu et al uses GRCh37, eQTL catalogue uses GRCh38
- Mu et al uses GENCODE v19, eQTL catalog uses GENCODE v30
- The Additional File 1 Table 8 says that the Zenodo files are: "List of significant eQTLs and sQTLs"